rust_snuba/src/processors/generic_metrics.rs

```diff
@@ -0,0 +1,1173 @@
+use adler32::adler32;
+use anyhow::Context;
+use chrono::DateTime;
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, vec};
+
+use crate::{
+    types::{InsertBatch, RowData},
+    KafkaMessageMetadata, ProcessorConfig,
+};
+
+use super::utils::enforce_retention;
+
+const GRANULARITY_TEN_SECONDS: u8 = 0;
+const GRANULARITY_ONE_MINUTE: u8 = 1;
+const GRANULARITY_ONE_HOUR: u8 = 2;
+const GRANULARITY_ONE_DAY: u8 = 3;
+
+/// Generate a timeseries ID from the given parameters. Timeseries IDs are used to
+/// uniquely identify a timeseries in the database. This implemenation is based on
+/// the python implementation in order to ensure that the same timeseries ID is
+/// generated for the same input.
+///
+/// Upstream sends the tag keys and values as strings. Which is surprising since the
+/// tag keys are indexed so they should be integers. But we have to deal with it.
+fn generate_timeseries_id(
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    tags: &BTreeMap<String, String>,
+) -> u32 {
+    let mut buffer = Vec::new();
+    buffer.extend_from_slice(&org_id.to_le_bytes());
+    buffer.extend_from_slice(&project_id.to_le_bytes());
+    buffer.extend_from_slice(&metric_id.to_le_bytes());
+
+    for (key, value) in tags {
+        buffer.extend_from_slice(key.as_bytes());
+        buffer.extend_from_slice(value.as_bytes());
+    }
+
+    adler32(&*buffer).unwrap()
+}
+
+#[derive(Debug, Deserialize)]
+struct FromGenericMetricsMessage {
+    // TODO: We don't seem to be performing any validation on the version. Should we?
```

Comment:
Nah, I think we are good. We check it's 2 in the schema https://github.com/getsentry/sentry-kafka-schemas/blob/main/schemas/snuba-generic-metrics.v1.schema.json#L10-L12, and I think this was mostly relevant in the past when we were migrating.